### ISSUE TITLE
key_value function ignoring whitespace again but matches on '?'

### ISSUE THREAD

**Issue by zez3** [Graylog2/graylog2-server#8048]
This bug is similar to https://github.com/Graylog2/graylog2-server/issues/4920

I have a set of data with key value pairs separated by a white space " ". One of my values contains the ? char.
The function instead of matching the white space it matches the char `'?'`

Example 
 `hostname="www.awin1.com" incidentserialno=965282358 url="/awclick.php?clickref=rcuserid=7315&mid=14110&id=403267" msg="Web.Client: HTTP.BROWSER_Firefox," apprisk="elevated"`

my pipeline rule:
```
rule "Parse connection messages"
when
    contains(to_string($message.message),"devid",true)
then
    let msg = key_value(
        value: to_string($message.message), allow_dup_keys:false, trim_value_chars: "\""
    );
    //delimiters	String		The characters used to separate pairs, defaults to whitespace
    //kv_delimiters	String		The characters used to separate keys from values, defaults to, defaults to '='
    set_fields(msg);
    remove_field("level");
end
```

## Expected Behavior
key_value function should not match the '?' character inside 
 url="awclick.php\?**clickref=rcuserid=7315&mid=14110&id=403267**"

## Current Behavior
key_value function is matching on `'?'` character and creates an field named key clickref with that `rcuserid=7315&mid=14110&id=403267` value

## Possible Solution
Fix inside this file:
https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/strings/KeyValue.java

Probably we need to escape `'?'` character

## Steps to Reproduce (for bugs)
try some logs that contain '?' followed by your kv_delimiter

To reproduce this I send the following hand generated syslog message:
`echo -n -e '<023>date=2020-05-12 time=15:33:15 devname="test" devid="12345" logid="1059028704" type="utm" subtype="app-ctrl" eventtype="signature" level="information" vd="test" eventtime=1587565996428508027 tz="+0200" appid=340r="test" group="TestGroup" authserver="mysrv.my.domain" srcip=172.x.x.x dstip=13.224.102.95 srcport=52916 dstport=80 srcintf="ssl.UNIVPN" srcintfrole="undefined" dstintf="OUT_749" dstintfrole="lan" proto=6 service="HTTP" direction="outgoing" policyid=14 sessionid=1665720512 applist="block_some" action="pass" appcat="Web.Client" app="HTTP.BROWSER_Firefox" hostname="btn.createsend1.com" incidentserialno=216809908 url="someStringTextHere+str_bfr=str_after.in.js?v=3" msg="Web.Client: HTTP.BROWSER_Firefox," app~risk="elevated" '"\0" | nc -w 1 -u mygraylog.my.domain 4522`

## Context
 This generates a lot of unneeded fields.

![image](https://user-images.githubusercontent.com/22619253/81784553-5318ea80-94fd-11ea-9258-eb83db94e459.png)

and a total of 8k fields
![image](https://user-images.githubusercontent.com/22619253/82341817-31ce6780-99f1-11ea-99f0-ada10451a776.png)

## Your Environment
* Graylog Version: 3.1.4
* Elasticsearch Version:  6.8.7-1
* MongoDB Version: 3.6
* Operating System: centos 7

**Comment by zez3** [comment]
Initially I thought that perhaps this happens because I did no explicitly set the delimiter to white space

> delimiters | String |   | The characters used to separate pairs, defaults to whitespace

but then I set it

> to_string($message.message),delimiters: " ", allow_dup_keys:false, trim_value_chars: "\""

so that did not helped but I just saw and it also matches on `'+'`
e.g.
url=""someStringTextHere+str_bfr=str_after.in.js?v=3"
will create an `str_bfr` field(key) with the value `str_after.in.js?v=3`

The `'?'` is ignored this time

@radykal-com 
Did you saw/observed this in your logs?

**Comment by zez3** [comment]
interesting enough, this:
`url="/MFYwVKADAgEAME0wSzBJMAkGBSsOAwIaBQAEFOe8Aj+c3j2o9MljapNC1+QvGO8QBBSjneYf+do5T8Bu6JHLlaXaMeIKnwIQCO8nQr+ofPm9VgGXILjr0Q=="`
will wrongly match only the last `'+'` creating the key `ofPm9VgGXILjr0Q` with the value `=`  

I have to check that guava CharMatcher, If someone is more knowledged, please help.

**Comment by zez3** [comment]
another interesting fact is that if I try to reproduce this bug in the UnitTests it passes
but in production is clearly failing

**Comment by bernd** [comment]
The described problem happened because the `key_value` function didn't support quoted values until 4.3. (see: https://github.com/Graylog2/graylog2-server/pull/11914)

The issue should be fixed with 4.3.

**Comment by zez3** [comment]
I needed this fixed 2-3 years ago :) I moved on
